src/modules/network/tcp/unix/unabto_tcp.c

```
@@ -0,0 +1,196 @@
+            return UTS_WOULD_BLOCK;
+        } else {
+            NABTO_LOG_ERROR(("unabto_tcp_read failed"));
+            return UTS_FAILED;
+        }
+    } else if (status == 0) {
+        NABTO_LOG_INFO(("TCP connection closed by peer"));
+        unabto_tcp_shutdown(sock);
+        unabto_tcp_close(sock);
+        return UTS_FAILED;
+    } else {
+        *read = status;
+        return UTS_OK;
+    }
+}
+
+unabto_tcp_status unabto_tcp_write(struct unabto_tcp_socket* sock, const void* buf, const size_t len, size_t* written){
+    int status;
+    NABTO_LOG_TRACE(("Writing %i bytes to tcp socket", len));
+    status = send(sock->socket, buf, len, MSG_NOSIGNAL);
+    NABTO_LOG_TRACE(("tcp send status: %i", status));
+    if (status > 0) {
+        *written = status;
+        return UTS_OK;
+    } else if (status < 0) {
+        int err = errno;
+        if ((err == EAGAIN) || err == EWOULDBLOCK) {
+            return UTS_WOULD_BLOCK;
+        } else {
+            NABTO_LOG_ERROR(("Send of tcp packet failed"));
+            unabto_tcp_close(sock);
+            return UTS_FAILED; 
+        }
+    }
+    return UTS_OK;
+}
+
+unabto_tcp_status unabto_tcp_close(struct unabto_tcp_socket* sock){
+    if (sock->socket == INVALID_SOCKET) {
+        NABTO_LOG_ERROR(("trying to close invalid socket"));
+    } else {
+#if NABTO_ENABLE_EPOLL
+        {
+            struct epoll_event ev;
+            memset(&ev, 0, sizeof(struct epoll_event));
+            epoll_ctl(unabto_epoll_fd, EPOLL_CTL_DEL, sock->socket, &ev);
+        }
+#endif
+        close(sock->socket);
+        sock->socket = INVALID_SOCKET;
+    }
+    return UTS_OK;
+}
+
+
+unabto_tcp_status unabto_tcp_shutdown(struct unabto_tcp_socket* sock){
+    shutdown(sock->socket, SHUT_WR);
+    return UTS_OK;
+}
+
+
+unabto_tcp_status unabto_tcp_open(struct unabto_tcp_socket* sock, void* dataPtr){
+    sock->socket = socket(AF_INET, SOCK_STREAM, 0);
+#if NABTO_ENABLE_EPOLL
```

maybe fail early here if (!socket) return fail.